File: backend/app/scrapers/leetcode.py
"""LeetCode statistics via the public GraphQL endpoint."""

from __future__ import annotations

import json
from datetime import datetime, timezone

import httpx

from ..schemas import (
    CodingBreakdownItem,
    CodingPlatformStats,
    CodingStatItem,
    Heatmap,
    HeatmapDay,
)
from .base import client, heatmap_ok, heatmap_unavailable, level_from_count, ok, unavailable

PLATFORM = "leetcode"
GRAPHQL = "https://leetcode.com/graphql"

CALENDAR_QUERY = """
query userCalendar($username: String!) {
  matchedUser(username: $username) {
    submissionCalendar
  }
}
"""

QUERY = """
query userStats($username: String!) {
  matchedUser(username: $username) {
    submitStatsGlobal {
      acSubmissionNum { difficulty count }
    }
  }
  userContestRanking(username: $username) {
    rating
    globalRanking
    attendedContestsCount
  }
}
"""


async def fetch(username: str) -> CodingPlatformStats:
    try:
        async with client(headers={"Content-Type": "application/json", "Referer": "https://leetcode.com"}) as http:
            resp = await http.post(
                GRAPHQL,
                json={"query": QUERY, "variables": {"username": username}},
            )
            resp.raise_for_status()
            payload = resp.json()

        data = payload.get("data") or {}
        matched = data.get("matchedUser")
        if not matched:
            return unavailable(PLATFORM, username, "Profile not found.")

        submissions = matched["submitStatsGlobal"]["acSubmissionNum"]
        by_diff = {row["difficulty"]: row["count"] for row in submissions}
        total = by_diff.get("All", 0)

        stats = [CodingStatItem(label="Solved", value=total)]
        contest = data.get("userContestRanking")
        if contest and contest.get("rating"):
            stats.append(CodingStatItem(label="Contest Rating", value=round(contest["rating"])))

        breakdown = [
            CodingBreakdownItem(name="Easy", value=by_diff.get("Easy", 0)),
            CodingBreakdownItem(name="Medium", value=by_diff.get("Medium", 0)),
            CodingBreakdownItem(name="Hard", value=by_diff.get("Hard", 0)),
        ]
        return ok(PLATFORM, username, stats, breakdown)
    except (httpx.HTTPError, ValueError, KeyError, TypeError):
        return unavailable(PLATFORM, username)


async def fetch_heatmap(username: str) -> Heatmap:
    """LeetCode submission calendar (public GraphQL). Returns a daily heatmap."""
    try:
        async with client(
            headers={"Content-Type": "application/json", "Referer": "https://leetcode.com"}
        ) as http:
            resp = await http.post(
                GRAPHQL,
                json={"query": CALENDAR_QUERY, "variables": {"username": username}},
            )
            resp.raise_for_status()
            payload = resp.json()

        matched = (payload.get("data") or {}).get("matchedUser")
        if not matched:
            return heatmap_unavailable(PLATFORM, username, "Profile not found.")

        raw_calendar = matched.get("submissionCalendar")
        if not raw_calendar:
            return heatmap_unavailable(PLATFORM, username, "No activity data available.")

        # submissionCalendar is a JSON string: {"<unix_seconds>": count, ...}
        calendar: dict[str, int] = json.loads(raw_calendar)
        days: list[HeatmapDay] = []
        for ts, count in calendar.items():
            date = datetime.fromtimestamp(int(ts), tz=timezone.utc).strftime("%Y-%m-%d")
            days.append(HeatmapDay(date=date, count=int(count), level=level_from_count(int(count))))

        if not days:
            return heatmap_unavailable(PLATFORM, username, "No activity data available.")
        days.sort(key=lambda d: d.date)
        return heatmap_ok(PLATFORM, username, days)
    except (httpx.HTTPError, ValueError, KeyError, TypeError, json.JSONDecodeError):
        return heatmap_unavailable(PLATFORM, username)

File: backend/src/scrapers/leetcode.ts
/** LeetCode statistics via the public GraphQL endpoint. */

import {
  heatmapOk,
  heatmapUnavailable,
  httpPost,
  levelFromCount,
  ok,
  unavailable,
} from './base'
import type { CodingBreakdownItem, CodingPlatformStats, CodingStatItem, Heatmap, HeatmapDay } from '../types'

const PLATFORM = 'leetcode'
const GRAPHQL = 'https://leetcode.com/graphql'

const CALENDAR_QUERY = `
query userCalendar($username: String!) {
  matchedUser(username: $username) {
    submissionCalendar
  }
}`

const QUERY = `
query userStats($username: String!) {
  matchedUser(username: $username) {
    submitStatsGlobal {
      acSubmissionNum { difficulty count }
    }
  }
  userContestRanking(username: $username) {
    rating
    globalRanking
    attendedContestsCount
  }
}`

const LC_HEADERS = { 'Content-Type': 'application/json', Referer: 'https://leetcode.com' }

export async function fetch_(username: string): Promise<CodingPlatformStats> {
  try {
    const resp = await httpPost(
      GRAPHQL,
      { query: QUERY, variables: { username } },
      { headers: LC_HEADERS },
    )
    if (!resp.ok) return unavailable(PLATFORM, username)
    const payload = (await resp.json()) as {
      data?: {
        matchedUser?: { submitStatsGlobal?: { acSubmissionNum?: { difficulty: string; count: number }[] } }
        userContestRanking?: { rating?: number } | null
      }
    }

    const data = payload.data ?? {}
    const matched = data.matchedUser
    if (!matched) return unavailable(PLATFORM, username, 'Profile not found.')

    const submissions = matched.submitStatsGlobal?.acSubmissionNum ?? []
    const byDiff = new Map(submissions.map((row) => [row.difficulty, row.count]))
    const total = byDiff.get('All') ?? 0

    const stats: CodingStatItem[] = [{ label: 'Solved', value: total }]
    const contest = data.userContestRanking
    if (contest && contest.rating) {
      stats.push({ label: 'Contest Rating', value: Math.round(contest.rating) })
    }

    const breakdown: CodingBreakdownItem[] = [
      { name: 'Easy', value: byDiff.get('Easy') ?? 0 },
      { name: 'Medium', value: byDiff.get('Medium') ?? 0 },
      { name: 'Hard', value: byDiff.get('Hard') ?? 0 },
    ]
    return ok(PLATFORM, username, stats, breakdown)
  } catch {
    return unavailable(PLATFORM, username)
  }
}

export async function fetchHeatmap(username: string): Promise<Heatmap> {
  try {
    const resp = await httpPost(
      GRAPHQL,
      { query: CALENDAR_QUERY, variables: { username } },
      { headers: LC_HEADERS },
    )
    if (!resp.ok) return heatmapUnavailable(PLATFORM, username)
    const payload = (await resp.json()) as {
      data?: { matchedUser?: { submissionCalendar?: string } }
    }

    const matched = payload.data?.matchedUser
    if (!matched) return heatmapUnavailable(PLATFORM, username, 'Profile not found.')

    const rawCalendar = matched.submissionCalendar
    if (!rawCalendar) return heatmapUnavailable(PLATFORM, username, 'No activity data available.')

    // submissionCalendar is a JSON string: {"<unix_seconds>": count, ...}
    const calendar = JSON.parse(rawCalendar) as Record<string, number>
    const days: HeatmapDay[] = []
    for (const [ts, count] of Object.entries(calendar)) {
      const date = new Date(Number.parseInt(ts, 10) * 1000).toISOString().slice(0, 10)
      const c = Number(count)
      days.push({ date, count: c, level: levelFromCount(c) })
    }

    if (days.length === 0) return heatmapUnavailable(PLATFORM, username, 'No activity data available.')
    days.sort((a, b) => a.date.localeCompare(b.date))
    return heatmapOk(PLATFORM, username, days)
  } catch {
    return heatmapUnavailable(PLATFORM, username)
  }
}
